### PR TITLE
JENKINS: check for config.h

### PR DESCRIPTION
**Pull request by hoopoepg**
- check for all .c files include config.h

### CODE REVIEW

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>UNKNOWN</b> on 5 workers (click for details)</summary>

**Note:** the logs will be deleted after 08-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/d88c95e6888e2324606f7d1c39d0449a/raw/90eb14dcba939c18058d9cfc1fa274336da55c17/_main) | :question: ABORTED
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/d88c95e6888e2324606f7d1c39d0449a/raw/59a77e1b670424d3cad06f0e68ba44f52fed7f61/hpc-arm-cavium-jenkins_W0) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/d88c95e6888e2324606f7d1c39d0449a/raw/1d6d7083eb46fc479199269a0b9932bbb983d518/hpc-arm-cavium-jenkins_W1) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/d88c95e6888e2324606f7d1c39d0449a/raw/be49399624ddc8d0e556adc318cad09dd7399c42/hpc-arm-cavium-jenkins_W2) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/d88c95e6888e2324606f7d1c39d0449a/raw/53a32935756a36397c1d8d07a7125534b2084716/hpc-arm-cavium-jenkins_W3) | :question: UNKNOWN

File: contrib/test_jenkins.sh [hunk 1]
```diff
@@ -659,6 +659,23 @@ check_make_distcheck() {
...
+	srcdir=$PWD/../src
+
+	# Check if all .c files include config.h
+	echo "==== Checking for config.h files in directory $srcdir ===="
+
+	if [ `find $srcdir -name \*.c |xargs grep -Le '#\s*include\s\+"config.h"'|wc -l` -eq 0 ];
+	then
+		echo echo "ok 1 - build successful " >> check_config_h.tap
+	else
+		echo Missing include config.h in files: `find $srcdir -name \*.c |xargs grep -Le '#\s*include\s\+"config.h"'`;
```
Comment (dmitrygx) [inline contrib/test_jenkins.sh:674]: can we cache the results of `find $srcdir -name \*.c |xargs grep -Le '#\s*include\s\+"config.h"'` command to the variable?
and then use it for the `if` statement and `echo`

Comment (hoopoepg) [reply contrib/test_jenkins.sh]: yep, why not... and use wc -w (word count) instead of -l (line count)

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>ABORTED</b> on 8 workers (click for details)</summary>

**Note:** the logs will be deleted after 08-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/461b4f737abec44dacfcfdedcd0c5db46d0883a8/_main) | :question: ABORTED
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/378e78e49bc9a891cb132f46c9bde79f2dc58915/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/0aebc7a823fba9531b92d3c98b0577c1631e5219/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/e8f62e82c4e5882de0bf2c924f37916c430d547b/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/85570bffb9354d1e67084f3f8583272f8643b1fd/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/ff9f253501caac2516a4d6707fee8fa1e5e04bcf/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/dd11a77c9ed8becf742f1fbbf488c8126c5ba749/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/0bc045a049ded0ee27246ba2669a99d8/raw/7aa450883d6c51fb9568f3a3a75a90eedc90fa48/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: contrib/test_jenkins.sh [hunk 2]
```diff
@@ -659,6 +659,25 @@ check_make_distcheck() {
...
 
+check_config_h() {
+	echo 1..1 > check_config_h.tap
+
+	srcdir=$PWD/../src
+
+	# Check if all .c files include config.h
+	echo "==== Checking for config.h files in directory $srcdir ===="
+
+	missing=`find $srcdir -name \*.c |xargs grep -Le '#\s*include\s\+"config.h"'`
```
Comment (brminich) [inline contrib/test_jenkins.sh:670]: minor space after `|`

File: contrib/test_jenkins.sh [hunk 3]
```diff
@@ -659,6 +659,25 @@ check_make_distcheck() {
...
+	
+	if [ `echo $missing|wc -w` -eq 0 ];
```
Comment (brminich) [inline contrib/test_jenkins.sh:672]: plz add spaces before and after `|`, like in other places

Comment (brminich) [inline contrib/test_jenkins.sh:672]: no need for `;`, because `then` is on the next line

File: contrib/test_jenkins.sh [hunk 4]
```diff
@@ -659,6 +659,25 @@ check_make_distcheck() {
...
+	then
+		echo echo "ok 1 - build successful " >> check_config_h.tap
+	else
+		echo Error: missing include config.h in files: $missing;
```
Comment (brminich) [inline contrib/test_jenkins.sh:676]: plz remove `;` and better wrap the string with `""`

File: contrib/test_jenkins.sh [hunk 5]
```diff
@@ -659,6 +659,25 @@ check_make_distcheck() {
...
+	srcdir=$PWD/../src
+
+	# Check if all .c files include config.h
+	echo "==== Checking for config.h files in directory $srcdir ===="
+
+	missing=`find $srcdir -name \*.c |xargs grep -Le '#\s*include\s\+"config.h"'`
+	
+	if [ `echo $missing|wc -w` -eq 0 ];
+	then
+		echo echo "ok 1 - build successful " >> check_config_h.tap
```
Comment (brminich) [inline contrib/test_jenkins.sh:674]: maybe check succesful?

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (hoopoepg) [reply contrib/test_jenkins.sh]: added space

File: contrib/test_jenkins.sh (hunk 3 above)
Comment (hoopoepg) [reply contrib/test_jenkins.sh]: added space

Comment (hoopoepg) [reply contrib/test_jenkins.sh]: removed ```;```

File: contrib/test_jenkins.sh (hunk 5 above)
Comment (hoopoepg) [reply contrib/test_jenkins.sh]: reworded

File: contrib/test_jenkins.sh (hunk 4 above)
Comment (hoopoepg) [reply contrib/test_jenkins.sh]: removed ```;``` and wrapped to ```""```

**Comment by hoopoepg** [comment]
@brminich @dmitrygx are you ok with latest changes?
I updated regexp - used perl notation because original regexp failed on grep version 2.5

File: contrib/test_jenkins.sh [hunk 6]
```diff
@@ -659,6 +659,25 @@ check_make_distcheck() {
...
 
+check_config_h() {
+	echo 1..1 > check_config_h.tap
+
+	srcdir=$PWD/../src
+
+	# Check if all .c files include config.h
+	echo "==== Checking for config.h files in directory $srcdir ===="
+
+	missing=`find $srcdir -name \*.c | xargs grep -LP '\#\s*include\s+"config.h"'`
```
Comment (brminich) [inline contrib/test_jenkins.sh:670]: ok, but can use simple grep, like `grep '\#include[ ]*\"config\.h\"'`

Comment (hoopoepg) [reply contrib/test_jenkins.sh:670]: what about
```
#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
```

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

File: contrib/test_jenkins.sh (hunk 6 above)
Comment (brminich) [reply contrib/test_jenkins.sh:670]: would  need to add `[ ]*` after `#` as well
never mind, you exp LGTM

Comment (yosefe) [reply contrib/test_jenkins.sh:670]: what about .cc files like in ucx_perftest?
also, do we expect this include from header files?
or SRC files must include config.h before including any other header file?

Comment (hoopoepg) [reply contrib/test_jenkins.sh:670]: > what about .cc files like in ucx_perftest?

hmmm, good question... actually perftest should not use internal UCX headers, but it uses. I can add evaluation for such files, but not sure it is correct

> also, do we expect this include from header files?
> or SRC files must include config.h before including any other header file?

source file (not header) must start with ```#include "config.h"```, internal include files - not mandatory (may/may not - both acceptable, imho)

Comment (yosefe) [reply contrib/test_jenkins.sh:670]: > hmmm, good question... actually perftest should not use internal UCX headers, but it uses. I can add evaluation for such files, but not sure it is correct

I think it's ok to use config.h

> source file (not header) must start with #include "config.h", internal include files - not mandatory (may/may not - both acceptable, imho)

Can we check that config.h is included first ?

Comment (hoopoepg) [reply contrib/test_jenkins.sh:670]: > I think it's ok to use config.h

ok, will update test

> Can we check that config.h is included first ?
the problem is not all include files should be started with config.h, for example external headers should not include it, that is why .c files should be started with config.h

**Comment by hoopoepg** [comment]
@yosefe more comments?

**Pull request merged**